File: treedex/loaders.py
"""Document loaders for multiple file formats.

Each loader returns a list of dicts: [{"page_num": int, "text": str, "token_count": int}]
matching the format used by pdf_parser.extract_pages().
"""

import os
import re
from html.parser import HTMLParser

import tiktoken

_enc = tiktoken.get_encoding("cl100k_base")


def _count_tokens(text: str) -> int:
    return len(_enc.encode(text))


def _text_to_pages(text: str, chars_per_page: int = 3000) -> list[dict]:
    """Split plain text into synthetic pages by character count."""
    pages = []
    for i in range(0, len(text), chars_per_page):
        chunk = text[i : i + chars_per_page]
        pages.append({
            "page_num": len(pages),
            "text": chunk,
            "token_count": _count_tokens(chunk),
        })
    return pages


class PDFLoader:
    """Load PDF files using PyMuPDF."""

    def __init__(self, extract_images: bool = False, detect_headings: bool = False):
        self.extract_images = extract_images
        self.detect_headings = detect_headings

    def load(self, path: str) -> list[dict]:
        from treedex.pdf_parser import extract_pages
        return extract_pages(
            path,
            extract_images=self.extract_images,
            detect_headings=self.detect_headings,
        )


class TextLoader:
    """Load plain text or markdown files."""

    def __init__(self, chars_per_page: int = 3000):
        self.chars_per_page = chars_per_page

    def load(self, path: str) -> list[dict]:
        with open(path, "r", encoding="utf-8") as f:
            text = f.read()
        return _text_to_pages(text, self.chars_per_page)


class _HTMLStripper(HTMLParser):
    """Simple HTML-to-text converter using stdlib."""

    def __init__(self):
        super().__init__()
        self._parts: list[str] = []
        self._skip = False

    def handle_starttag(self, tag, attrs):
        if tag in ("script", "style"):
            self._skip = True
        if tag == "img":
            attrs_dict = dict(attrs)
            alt = attrs_dict.get("alt", "").strip()
            if alt:
                self._parts.append(f"\n[Image: {alt}]\n")
            else:
                self._parts.append("\n[Image]\n")

    def handle_endtag(self, tag):
        if tag in ("script", "style"):
            self._skip = False
        if tag in ("p", "div", "br", "h1", "h2", "h3", "h4", "h5", "h6", "li", "tr"):
            self._parts.append("\n")

    def handle_data(self, data):
        if not self._skip:
            self._parts.append(data)

    def get_text(self) -> str:
        raw = "".join(self._parts)
        return re.sub(r"\n{3,}", "\n\n", raw).strip()


class HTMLLoader:
    """Load HTML files, stripping tags to plain text (stdlib only)."""

    def __init__(self, chars_per_page: int = 3000):
        self.chars_per_page = chars_per_page

    def load(self, path: str) -> list[dict]:
        with open(path, "r", encoding="utf-8") as f:
            html = f.read()
        stripper = _HTMLStripper()
        stripper.feed(html)
        text = stripper.get_text()
        return _text_to_pages(text, self.chars_per_page)


class DOCXLoader:
    """Load DOCX files using python-docx."""

    def __init__(self, chars_per_page: int = 3000):
        self.chars_per_page = chars_per_page

    def load(self, path: str) -> list[dict]:
        import docx
        from docx.oxml.ns import qn

        doc = docx.Document(path)
        parts = []
        for paragraph in doc.paragraphs:
            parts.append(paragraph.text)
            # Check for inline images in the paragraph's XML
            for drawing in paragraph._element.findall(f".//{qn('wp:inline')}"):
                doc_pr = drawing.find(qn("wp:docPr"))
                if doc_pr is not None:
                    alt = doc_pr.get("descr", "").strip()
                    if alt:
                        parts.append(f"[Image: {alt}]")
                    else:
                        parts.append("[Image]")
        text = "\n".join(parts)
        return _text_to_pages(text, self.chars_per_page)


_EXTENSION_MAP = {
    ".pdf": PDFLoader,
    ".txt": TextLoader,
    ".md": TextLoader,
    ".html": HTMLLoader,
    ".htm": HTMLLoader,
    ".docx": DOCXLoader,
}


def auto_loader(
    path: str,
    extract_images: bool = False,
    detect_headings: bool = False,
) -> list[dict]:
    """Auto-detect file format and load pages."""
    ext = os.path.splitext(path)[1].lower()
    loader_cls = _EXTENSION_MAP.get(ext)
    if loader_cls is None:
        raise ValueError(
            f"Unsupported file extension '{ext}'. "
            f"Supported: {', '.join(_EXTENSION_MAP)}"
        )
    if ext == ".pdf":
        return PDFLoader(
            extract_images=extract_images,
            detect_headings=detect_headings,
        ).load(path)
    return loader_cls().load(path)

File: src/loaders.ts
/**
 * Document loaders for multiple file formats.
 *
 * Each loader returns a list of objects: [{page_num, text, token_count}]
 * matching the format used by pdf-parser extractPages().
 */

import { countTokens } from "./pdf-parser.js";
import type { Page } from "./types.js";

/** Split plain text into synthetic pages by character count. */
export function textToPages(
  text: string,
  charsPerPage: number = 3000,
): Page[] {
  const pages: Page[] = [];
  for (let i = 0; i < text.length; i += charsPerPage) {
    const chunk = text.slice(i, i + charsPerPage);
    pages.push({
      page_num: pages.length,
      text: chunk,
      token_count: countTokens(chunk),
    });
  }
  return pages;
}

/** Load PDF files using pdfjs-dist. */
export class PDFLoader {
  readonly extractImages: boolean;

  constructor(options?: { extractImages?: boolean }) {
    this.extractImages = options?.extractImages ?? false;
  }

  async load(path: string): Promise<Page[]> {
    const { extractPages } = await import("./pdf-parser.js");
    return extractPages(path, { extractImages: this.extractImages });
  }
}

/** Load plain text or markdown files. */
export class TextLoader {
  readonly charsPerPage: number;

  constructor(charsPerPage: number = 3000) {
    this.charsPerPage = charsPerPage;
  }

  async load(path: string): Promise<Page[]> {
    const fs = await import("node:fs/promises");
    const text = await fs.readFile(path, "utf-8");
    return textToPages(text, this.charsPerPage);
  }
}

/** Load HTML files, stripping tags to plain text. */
export class HTMLLoader {
  readonly charsPerPage: number;

  constructor(charsPerPage: number = 3000) {
    this.charsPerPage = charsPerPage;
  }

  async load(path: string): Promise<Page[]> {
    const fs = await import("node:fs/promises");
    const html = await fs.readFile(path, "utf-8");
    const text = await this.stripHtml(html);
    return textToPages(text, this.charsPerPage);
  }

  private async stripHtml(html: string): Promise<string> {
    try {
      // Try htmlparser2 if available
      // @ts-expect-error -- optional peer dependency
      const { Parser } = await import("htmlparser2");
      return new Promise((resolve) => {
        const parts: string[] = [];
        let skip = false;

        const parser = new Parser({
          onopentag(name: string, attribs: Record<string, string>) {
            if (name === "script" || name === "style") skip = true;
            if (name === "img") {
              const alt = (attribs.alt || "").trim();
              if (alt) {
                parts.push(`\n[Image: ${alt}]\n`);
              } else {
                parts.push("\n[Image]\n");
              }
            }
          },
          onclosetag(name: string) {
            if (name === "script" || name === "style") skip = false;
            if (
              ["p", "div", "br", "h1", "h2", "h3", "h4", "h5", "h6", "li", "tr"].includes(name)
            ) {
              parts.push("\n");
            }
          },
          ontext(data: string) {
            if (!skip) parts.push(data);
          },
          onend() {
            const raw = parts.join("");
            resolve(raw.replace(/\n{3,}/g, "\n\n").trim());
          },
        });

        parser.write(html);
        parser.end();
      });
    } catch {
      // Fallback: simple regex-based tag stripping
      // Extract img alt text before stripping all tags
      let processed = html
        .replace(/<script[^>]*>[\s\S]*?<\/script>/gi, "")
        .replace(/<style[^>]*>[\s\S]*?<\/style>/gi, "");
      processed = processed.replace(/<img\b[^>]*>/gi, (tag) => {
        const altMatch = tag.match(/alt=["']([^"']*)["']/i);
        const alt = altMatch ? altMatch[1].trim() : "";
        return alt ? ` [Image: ${alt}] ` : " [Image] ";
      });
      return processed
        .replace(/<[^>]+>/g, " ")
        .replace(/\s+/g, " ")
        .trim();
    }
  }
}

/** Load DOCX files using mammoth. */
export class DOCXLoader {
  readonly charsPerPage: number;

  constructor(charsPerPage: number = 3000) {
    this.charsPerPage = charsPerPage;
  }

  async load(path: string): Promise<Page[]> {
    const fs = await import("node:fs/promises");
    // @ts-expect-error -- optional peer dependency
    const mammoth = await import("mammoth");
    const buffer = await fs.readFile(path);
    const result = await mammoth.convertToHtml({ buffer });
    // Replace <img> tags with [Image: alt] markers, then strip remaining HTML
    let html: string = result.value;
    html = html.replace(/<img\b[^>]*>/gi, (tag: string) => {
      const altMatch = tag.match(/alt=["']([^"']*)["']/i);
      const alt = altMatch ? altMatch[1].trim() : "";
      return alt ? `[Image: ${alt}]` : "[Image]";
    });
    const text = html.replace(/<[^>]+>/g, " ").replace(/\s+/g, " ").trim();
    return textToPages(text, this.charsPerPage);
  }
}

interface Loader {
  load(path: string): Promise<Page[]>;
}

const EXTENSION_MAP: Record<string, { new (): Loader }> = {
  ".pdf": PDFLoader,
  ".txt": TextLoader,
  ".md": TextLoader,
  ".html": HTMLLoader,
  ".htm": HTMLLoader,
  ".docx": DOCXLoader,
};

/** Auto-detect file format and load pages. */
export async function autoLoader(
  filePath: string,
  options?: { extractImages?: boolean },
): Promise<Page[]> {
  const { extname } = await import("node:path");
  const ext = extname(filePath).toLowerCase();
  const LoaderClass = EXTENSION_MAP[ext];
  if (!LoaderClass) {
    const supported = Object.keys(EXTENSION_MAP).join(", ");
    throw new Error(
      `Unsupported file extension '${ext}'. Supported: ${supported}`,
    );
  }
  if (ext === ".pdf" && options?.extractImages) {
    return new PDFLoader({ extractImages: true }).load(filePath);
  }
  const loader = new LoaderClass();
  return loader.load(filePath);
}
